numpy/core/numeric.py

```diff
@@ -376,6 +376,153 @@ def extend_all(module):
+
+    if not axis and axis != 0:
+        axis = None
+
+    # multiarray is faster with
+    # booleans when axis is None
+    if a.dtype == bool:
+        return (multiarray.count_nonzero(a) if axis is None
+                else a.sum(axis=axis, dtype=intp))
+
+    # sum is faster with numbers
+    # independent of the axis
+    if issubdtype(a.dtype, number):
+        return (a != 0).sum(axis=axis, dtype=intp)
+
+    # multiarray is faster with
+    # strings when axis is None
+    if issubdtype(a.dtype, string_) or issubdtype(a.dtype, unicode_):
+        nullstr = a.dtype.type('')
+        return (multiarray.count_nonzero(a) if axis is None
+                else (a != nullstr).sum(axis=axis, dtype=intp))
+
+    if axis is None:
+        return multiarray.count_nonzero(a)
+
+    axis = atleast_1d(axis)
+
+    if ((axis.ndim > 1 and axis.shape[0] != 1) or
+            not issubdtype(axis.dtype, number)):
+        raise ValueError(('Invalid axis provided. Cannot '
+                          'convert to scalar index.'), axis)
+
+    if unique(axis).size != axis.size:
+        raise ValueError("Duplicate value in 'axis'")
+
+    axis = axis.flatten()
+    counts = apply_along_axis(multiarray.count_nonzero, axis[0], a)
+
+    if axis.size == 1:
+        return counts
+    else:
+        # for subsequent axis numbers, that number decreases
+        # by one in this new 'counts' array if it was larger
+        # than the first axis upon which 'count_nonzero' was
+        # applied but remains unchanged if that number was
+        # smaller than that first axis
+        #
+        # this trick enables us to perform counts on object-like
+        # elements across multiple axes very quickly because integer
+        # addition is very well optimized
+        return counts.sum(axis=tuple(axis[1:] - (
+            axis[1:] > axis[0])), dtype=intp)
+
+newaxis = None
+
+arange = multiarray.arange
+array = multiarray.array
+zeros = multiarray.zeros
+empty = multiarray.empty
+empty_like = multiarray.empty_like
+fromstring = multiarray.fromstring
+fromiter = multiarray.fromiter
+fromfile = multiarray.fromfile
+frombuffer = multiarray.frombuffer
+shares_memory = multiarray.shares_memory
+may_share_memory = multiarray.may_share_memory
+if sys.version_info[0] < 3:
+    newbuffer = multiarray.newbuffer
+    getbuffer = multiarray.getbuffer
+int_asbuffer = multiarray.int_asbuffer
+where = multiarray.where
+concatenate = multiarray.concatenate
+fastCopyAndTranspose = multiarray._fastCopyAndTranspose
+set_numeric_ops = multiarray.set_numeric_ops
+can_cast = multiarray.can_cast
+promote_types = multiarray.promote_types
+min_scalar_type = multiarray.min_scalar_type
+result_type = multiarray.result_type
+lexsort = multiarray.lexsort
+compare_chararrays = multiarray.compare_chararrays
+putmask = multiarray.putmask
+einsum = multiarray.einsum
+dot = multiarray.dot
+inner = multiarray.inner
+vdot = multiarray.vdot
+matmul = multiarray.matmul
```

Comment:
Why are all these needed? Looks unrelated to the rest of this PR and not desired.